File: runtimes/android/host/src/main/kotlin/dev/flux/host/vm/debug/Telemetry.kt
package dev.flux.host.vm.debug

import dev.flux.host.vm.FluxValue
import java.io.ByteArrayOutputStream

/**
 * DevTools bidirectional debug telemetry (spec §3, §4).
 *
 * Host-side mirror of `flux_ir_serde::TelemetryEvent` / `DebugCommand` (Appendix
 * D §D.12, ADR-0039). The byte layout produced by [TelemetryEvent.toFrameBytes]
 * is bit-identical to `crates/flux-ir-serde/src/telemetry.rs` so the Rust dev
 * server decodes host frames without translation.
 *
 * Every call site that emits is guarded by `BuildConfig.DEBUG` (spec §3 Key
 * Principle 1: zero release impact).
 */

/** A debug telemetry event emitted by the host runtime (raw, host → server). */
public sealed interface TelemetryEvent {
    /** Emitted after each VM instruction executes. */
    public data class VmStep(
        val bytecodeOffset: UInt,
        val opcode: UByte,
        val registers: List<FluxValue>,
        val gasRemaining: UInt,
    ) : TelemetryEvent

    /** Emitted when a signal's value changes. */
    public data class SignalWrite(
        val signalId: UInt,
        val oldValue: FluxValue,
        val newValue: FluxValue,
        val triggeredEffectIds: List<UInt>,
    ) : TelemetryEvent

    /** Emitted when the reconciler mutates a native view. */
    public data class ViewMutation(
        val nodeId: UInt,
        val nativeViewId: ULong,
        val mutationKind: UByte,
        val frame: Rect?,
    ) : TelemetryEvent

    /** Emitted when a handler starts or finishes. */
    public data class HandlerInvocation(
        val handlerId: UInt,
        val isStart: Boolean,
        val gasUsed: UInt?,
    ) : TelemetryEvent
}

/** A native view layout rectangle (device points). */
public data class Rect(
    val x: Double,
    val y: Double,
    val width: Double,
    val height: Double,
)

/** A control command sent DevTools → host (Appendix D §D.12 §2.3). */
public sealed interface DebugCommand {
    public data object Pause : DebugCommand

    public data object Resume : DebugCommand

    public data object Step : DebugCommand

    public data class SetBreakpoint(
        val bytecodeOffset: UInt,
    ) : DebugCommand

    public data class ClearBreakpoint(
        val bytecodeOffset: UInt,
    ) : DebugCommand

    public data object RequestSnapshot : DebugCommand
}

/** The `MAGIC` header for every Flux wire frame (Appendix D §D.1). */
internal const val MAGIC: UInt = 0x465C5558u

internal const val FRAME_TELEMETRY: UByte = 0x10u
private const val FRAME_DEBUG_COMMAND: UByte = 0x11u

/** Encodes a [FluxValue] into [out] per Appendix D §D.5. */
private fun encodeValue(
    value: FluxValue,
    out: ByteArrayOutputStream,
) {
    when (value) {
        FluxValue.NullVal -> out.write(0x00)
        is FluxValue.IntVal -> {
            out.write(0x01)
            out.writeLongLE(value.value)
        }
        is FluxValue.FloatVal -> {
            out.write(0x02)
            out.writeLongLE(value.value.toBits())
        }
        is FluxValue.BoolVal -> {
            out.write(0x03)
            out.write(if (value.value) 1 else 0)
        }
        is FluxValue.StrVal -> {
            out.write(0x04)
            out.writeUIntLE(value.id)
        }
        is FluxValue.HandlerRefVal -> {
            out.write(0x05)
            out.writeUIntLE(value.handlerId)
        }
        is FluxValue.ListVal -> {
            out.write(0x06)
            out.writeUShortLE(value.items.size.toUShort())
            for (item in value.items) encodeValue(item, out)
        }
        is FluxValue.RecordVal -> {
            out.write(0x07)
            out.writeUShortLE(value.fields.size.toUShort())
            for (field in value.fields) {
                out.writeUShortLE(field.index)
                encodeValue(field.value, out)
            }
        }
    }
}

private fun writeUIntLE0(
    out: ByteArrayOutputStream,
    v: UInt,
) = out.write(v.toInt())

private fun writeUShortLE0(
    out: ByteArrayOutputStream,
    v: UShort,
) {
    out.write((v.toInt() and 0xFF))
    out.write((v.toInt() ushr 8) and 0xFF)
}

private fun writeLongLE0(
    out: ByteArrayOutputStream,
    v: Long,
) {
    var x = v
    repeat(8) {
        out.write((x.toInt() and 0xFF))
        x = x shr 8
    }
}

private fun ByteArrayOutputStream.writeUIntLE(v: UInt) = writeUIntLE0(this, v)

private fun ByteArrayOutputStream.writeUShortLE(v: UShort) = writeUShortLE0(this, v)

private fun ByteArrayOutputStream.writeLongLE(v: Long) = writeLongLE0(this, v)

/** Encodes this event as a length-prefixed union body (no frame header). */
internal fun TelemetryEvent.encodeBody(out: ByteArrayOutputStream) {
    val body = ByteArrayOutputStream()
    when (this) {
        is TelemetryEvent.VmStep -> {
            body.write(0x01)
            body.writeUIntLE(bytecodeOffset)
            body.write(opcode.toInt())
            for (reg in registers.take(16)) encodeValue(reg, body)
            body.writeUIntLE(gasRemaining)
        }
        is TelemetryEvent.SignalWrite -> {
            body.write(0x02)
            body.writeUIntLE(signalId)
            encodeValue(oldValue, body)
            encodeValue(newValue, body)
            body.writeUShortLE(triggeredEffectIds.size.toUShort())
            for (effect in triggeredEffectIds) body.writeUIntLE(effect)
        }
        is TelemetryEvent.ViewMutation -> {
            body.write(0x03)
            body.writeUIntLE(nodeId)
            writeULongLE(body, nativeViewId)
            body.write(mutationKind.toInt())
            if (frame != null) {
                body.write(0x01)
                writeDoubleBitsLE(body, frame.x)
                writeDoubleBitsLE(body, frame.y)
                writeDoubleBitsLE(body, frame.width)
                writeDoubleBitsLE(body, frame.height)
            } else {
                body.write(0x00)
            }
        }
        is TelemetryEvent.HandlerInvocation -> {
            body.write(0x04)
            body.writeUIntLE(handlerId)
            body.write(if (isStart) 1 else 0)
            if (gasUsed != null) {
                body.write(0x01)
                body.writeUIntLE(gasUsed)
            } else {
                body.write(0x00)
            }
        }
    }
    // Length-prefix (u32 LE) the body, back-patched at the front.
    val bytes = body.toByteArray()
    writeUIntLE0(out, bytes.size.toUInt())
    out.write(bytes)
}

private fun writeULongLE(
    out: ByteArrayOutputStream,
    v: ULong,
) {
    var x = v
    repeat(8) {
        out.write((x.toInt() and 0xFF))
        x = x shr 8
    }
}

private fun writeDoubleBitsLE(
    out: ByteArrayOutputStream,
    v: Double,
) = writeLongLE0(out, v.toBits())

/** Encodes this event into a full `Telemetry` frame (Appendix D §D.12). */
public fun TelemetryEvent.toFrameBytes(version: UByte = 0x01u): ByteArray {
    val out = ByteArrayOutputStream()
    out.writeUIntLE(MAGIC)
    out.write(version.toInt())
    out.write(FRAME_TELEMETRY.toInt())
    out.writeUShortLE(1u)
    this.encodeBody(out)
    return out.toByteArray()
}

/** Encodes this command into a full `DebugCommand` frame (kind `0x11`). */
public fun DebugCommand.toFrameBytes(
    commandId: UInt,
    version: UByte = 0x01u,
): ByteArray {
    val payload = ByteArrayOutputStream()
    when (this) {
        DebugCommand.Pause -> payload.write(0x01)
        DebugCommand.Resume -> payload.write(0x02)
        DebugCommand.Step -> payload.write(0x03)
        is DebugCommand.SetBreakpoint -> {
            payload.write(0x04)
            payload.writeUIntLE(bytecodeOffset)
        }
        is DebugCommand.ClearBreakpoint -> {
            payload.write(0x05)
            payload.writeUIntLE(bytecodeOffset)
        }
        DebugCommand.RequestSnapshot -> payload.write(0x06)
    }
    val out = ByteArrayOutputStream()
    out.writeUIntLE(MAGIC)
    out.write(version.toInt())
    out.write(FRAME_DEBUG_COMMAND.toInt())
    out.writeUIntLE(commandId)
    out.writeUShortLE(payload.size().toUShort())
    out.write(payload.toByteArray())
    return out.toByteArray()
}

File: runtimes/ios/FluxHost/Sources/FluxHost/Debug/Telemetry.swift
//  Telemetry.swift
//  DevTools bidirectional debug telemetry (spec §3, §4).
//
//  Defines the host-side `TelemetryEvent` / `DebugCommand` model and a
//  thread-safe `TelemetryBridge` that batches events and encodes them into the
//  `Telemetry` (`0x10`) / `DebugCommand` (`0x11`) frames defined by
//  `flux-ir-serde` (Appendix D §D.12, ADR-0039). The byte layout here is kept
//  bit-identical to `crates/flux-ir-serde/src/telemetry.rs` so the Rust dev
//  server decodes host frames without a translation step.
//
//  ┌───────────────────────────────────────────────────────────────────────┐
//  │ RELEASE COMPILE-OUT (brittleness 8c). The entire telemetry/"trace" sink  │
//  │ is wrapped in `#if DEBUG`. In a Release build none of this — the        │
//  │ `TelemetryEvent`/`DebugCommand` models, the `TelemetryBridge`, the       │
//  │ `DevToolsSocket`, the `encodeValue` helper, nor the `fluxDevtoolsEmit`   │
//  │ no-op — is compiled or linked. The VM (`FluxBytecodeVM`) and signal graph│
//  │ (`SignalGraph`) emit only inside their own `#if DEBUG` blocks, which now │
//  │ reference a symbol that does not exist in Release, so the optimiser       │
//  │ drops every trace call site and leaves zero dead telemetry code linked.  │
//  └───────────────────────────────────────────────────────────────────────┘

import Foundation
import os.log

/// A native view layout rectangle (device points). Mirrors the wire `Rect`.
public struct Rect: Sendable, Equatable {
    /// Left edge (x origin).
    public let x: Double
    /// Top edge (y origin).
    public let y: Double
    /// Width.
    public let width: Double
    /// Height.
    public let height: Double

    public init(x: Double, y: Double, width: Double, height: Double) {
        self.x = x
        self.y = y
        self.width = width
        self.height = height
    }
}

/// Protocol a DevTools transport conforms to in order to receive VM telemetry.
public protocol VMTelemetrySink: AnyObject {
    /// Receives one decoded telemetry event from the host VM / signal graph.
    func emit(_ event: TelemetryEvent)
}

/// A debug telemetry event emitted by the host runtime (raw, host → server).
///
/// Mirrors `flux_ir_serde::TelemetryEvent` (Appendix D §D.12 §2.2). Source-span
/// enrichment happens server-side, so these carry raw IDs only.
public enum TelemetryEvent {
    /// Emitted after each VM instruction executes.
    case vmStep(bytecodeOffset: UInt32, opcode: UInt8, registers: [FluxValue], gasRemaining: UInt32)
    /// Emitted when a signal's value changes.
    case signalWrite(signalId: UInt32, oldValue: FluxValue, newValue: FluxValue, triggeredEffectIds: [UInt32])
    /// Emitted when the reconciler mutates a native view.
    case viewMutation(nodeId: UInt32, nativeViewId: UInt64, parentId: UInt32, mutationKind: UInt8, frame: Rect?)
    /// Emitted when a handler starts or finishes.
    case handlerInvocation(handlerId: UInt32, isStart: Bool, gasUsed: UInt32?)

    /// Encodes this event as a length-prefixed union body (no frame header).
    func encode(into data: inout Data) {
        let start = data.count
        // Reserve a 4-byte length slot; back-patch after the body is written.
        data.append(contentsOf: [UInt8](repeating: 0, count: 4))
        switch self {
        case let .vmStep(offset, opcode, registers, gas):
            data.append(0x01)
            data.append(contentsOf: offset.bytesLE())
            data.append(opcode)
            for reg in registers.prefix(16) {
                encodeValue(reg, into: &data)
            }
            data.append(contentsOf: gas.bytesLE())
        case let .signalWrite(id, oldV, newV, effects):
            data.append(0x02)
            data.append(contentsOf: id.bytesLE())
            encodeValue(oldV, into: &data)
            encodeValue(newV, into: &data)
            data.append(contentsOf: UInt16(effects.count).bytesLE())
            for effect in effects {
                data.append(contentsOf: effect.bytesLE())
            }
        case let .viewMutation(node, native, parent, kind, frame):
            data.append(0x03)
            data.append(contentsOf: node.bytesLE())
            data.append(contentsOf: native.bytesLE())
            data.append(contentsOf: parent.bytesLE())
            data.append(kind)
            if let rect = frame {
                data.append(0x01)
                data.append(contentsOf: rect.x.bitPattern.bytesLE())
                data.append(contentsOf: rect.y.bitPattern.bytesLE())
                data.append(contentsOf: rect.width.bitPattern.bytesLE())
                data.append(contentsOf: rect.height.bitPattern.bytesLE())
            } else {
                data.append(0x00)
            }
        case let .handlerInvocation(handler, isStart, gas):
            data.append(0x04)
            data.append(contentsOf: handler.bytesLE())
            data.append(isStart ? 1 : 0)
            if let g = gas {
                data.append(0x01)
                data.append(contentsOf: g.bytesLE())
            } else {
                data.append(0x00)
            }
        }
        let len = UInt32(data.count - start - 4)
        let lenBytes = len.bytesLE()
        data.replaceSubrange(start ..< start + 4, with: lenBytes)
    }

    /// Encodes this event into a full `Telemetry` frame (Appendix D §D.12).
    func toFrameBytes(version: UInt8 = 1) -> Data {
        var data = Data()
        data.append(contentsOf: FluxTelemetryMagic.bytesLE())
        data.append(version)
        data.append(0x10) // FRAME_TELEMETRY
        data.append(contentsOf: UInt16(1).bytesLE()) // event_count
        self.encode(into: &data)
        return data
    }
}

/// The `MAGIC` header for every Flux wire frame (Appendix D §D.1).
let FluxTelemetryMagic: UInt32 = 0x465C_5558

/// A control command sent DevTools → host (Appendix D §D.12 §2.3).
public enum DebugCommand {
    case pause
    case resume
    case step
    case setBreakpoint(bytecodeOffset: UInt32)
    case clearBreakpoint(bytecodeOffset: UInt32)
    case requestSnapshot

    /// Encodes this command into a full `DebugCommand` frame (kind `0x11`).
    func toFrameBytes(commandId: UInt32, version: UInt8 = 1) -> Data {
        var payload = Data()
        switch self {
        case .pause: payload.append(0x01)
        case .resume: payload.append(0x02)
        case .step: payload.append(0x03)
        case let .setBreakpoint(offset):
            payload.append(0x04)
            payload.append(contentsOf: offset.bytesLE())
        case let .clearBreakpoint(offset):
            payload.append(0x05)
            payload.append(contentsOf: offset.bytesLE())
        case .requestSnapshot: payload.append(0x06)
        }
        var data = Data()
        data.append(contentsOf: FluxTelemetryMagic.bytesLE())
        data.append(version)
        data.append(0x11) // FRAME_DEBUG_COMMAND
        data.append(contentsOf: commandId.bytesLE())
        data.append(contentsOf: UInt16(payload.count).bytesLE())
        data.append(payload)
        return data
    }
}

/// Thread-safe batching queue that flushes telemetry to the WebSocket.
///
/// The VM and signal graph call `emit` from their evaluation context; the
/// bridge serializes onto its own queue and batches (every 16 ms or on a
/// 10-event threshold) so instrumentation never blocks the VM or UI thread
/// (spec §3.3).
public final class TelemetryBridge: VMTelemetrySink {
    /// Flush once this many events accumulate in the batch.
    private static let batchSize = 10
    /// Maximum time an event waits in the batch before a flush is forced.
    private static let batchInterval: TimeInterval = 0.016
    private let queue = DispatchQueue(label: "dev.flux.telemetry")
    private var pending: [TelemetryEvent] = []
    /// A pending 16 ms flush timer; cancelled when a threshold flush fires first.
    private var flushTimer: DispatchWorkItem?

    /// Appends an event to the batch and flushes (a) once `batchSize` events
    /// accumulate or (b) after `batchInterval` elapses, whichever comes first.
    /// Batching keeps instrumentation off the VM/UI hot path even when `vmStep`
    /// emits one event per instruction (spec §3.3). A single `signalWrite` (e.g.
    /// one counter tap) still drains within 16 ms, so the live path stays
    /// observable. (Restores the batching that the per-event `flush()` DIAGNOSIS
    /// temporarily disabled.)
    public func emit(_ event: TelemetryEvent) {
        queue.async {
            self.pending.append(event)
            if self.pending.count >= TelemetryBridge.batchSize {
                self.flushTimer?.cancel()
                self.flushTimer = nil
                self.flush()
                return
            }
            if self.flushTimer == nil {
                let item = DispatchWorkItem { [weak self] in
                    self?.flushTimer = nil
                    self?.flush()
                }
                self.flushTimer = item
                self.queue.asyncAfter(deadline: .now() + TelemetryBridge.batchInterval, execute: item)
            }
        }
    }

    /// Encodes and forwards all pending events (called on the bridge queue).
    func flush() {
        guard !pending.isEmpty else { return }
        let batch = pending
        pending.removeAll()
        var frame = Data()
        frame.append(contentsOf: FluxTelemetryMagic.bytesLE())
        frame.append(1)
        frame.append(0x10)
        frame.append(contentsOf: UInt16(batch.count).bytesLE())
        for event in batch {
            event.encode(into: &frame)
        }
        // The concrete send is performed by the host transport, which retains
        // this bridge and observes `pending` via `takePending()`.
        onFlush?(frame)
    }

    /// Hook the host transport sets to actually transmit a batched frame.
    var onFlush: ((Data) -> Void)?

    /// Drains and returns all pending events (test/transport helper).
    func takePending() -> [TelemetryEvent] {
        queue.sync {
            let batch = pending
            pending.removeAll()
            return batch
        }
    }
}

/// The active DevTools telemetry sink, or `nil` when DevTools is disconnected.
///
/// The VM and signal graph call `fluxDevtoolsEmit` after each observable step;
/// when no sink is attached the call is a no-op (zero release impact).
///
/// Marked `nonisolated(unsafe)` because it is externally synchronized: it is
/// assigned exactly once at host startup (before any handler runs) and only
/// ever read afterwards, so the Swift 6 concurrency checker's global-mutable
/// state rule does not apply.
nonisolated(unsafe) public var fluxDevtoolsSink: (any VMTelemetrySink)?

/// Safe append-only diagnostic logger to the app's Documents directory.
/// The simulator's container lives on the Mac filesystem, so this is the
/// reliable way to observe package-target (FluxHost) behaviour from the host.
func fluxTrace(_ line: String) {
    let fm = FileManager.default
    guard let dir = fm.urls(for: .documentDirectory, in: .userDomainMask).first else { return }
    let f = dir.appendingPathComponent("flux_tele.log")
    let stamped = "\(Date()) \(line)\n"
    if let fh = try? FileHandle(forWritingTo: f) {
        fh.seekToEndOfFile()
        fh.write(stamped.data(using: .utf8)!)
        try? fh.close()
    } else {
        try? stamped.data(using: .utf8)?.write(to: f)
    }
}

/// Opens the host → DevTools channel and installs the telemetry sink so
/// emitted VM/signal events flow to the dev server. Telemetry is sent over the
/// existing host patch-channel WebSocket (`send`, supplied by the app — it is
/// the same `FluxTransport.send` the host already uses for dispatch events).
/// Routing to connected DevTools apps happens server-side (the dev server fans
/// `Telemetry` `0x10` frames received on `:7331` out to every DevTools
/// subscriber). This avoids a separate device→:7333 socket, which the iOS
/// Simulator cannot reach (it only forwards the `:7331` loopback port).
///
/// Safe to call when no dev server is running: `send` simply drops frames.
public func fluxDevtoolsConnect(send: @escaping (Data) -> Void) {
    os_log(.fault, "RTFLUXTELE CONNECT_CALLED")
    let bridge = TelemetryBridge()
    fluxDevtoolsSetSink(bridge)
    bridge.onFlush = { frame in
        send(frame)
    }
    // Replay the host's current shadow tree so a freshly-connected DevTools
    // shows the component hierarchy immediately, without waiting for the next
    // mount/tap. The reconciler registers this hook at startup.
    fluxDevtoolsOnConnect?()
}

/// Hook the host sets to replay its current shadow tree when DevTools connects.
///
/// Called exactly once, right after the telemetry sink is attached, so the
/// DevTools component tree populates from the already-built node graph instead
/// of requiring a fresh reconcile (which only happens on mount or interaction).
nonisolated(unsafe) public var fluxDevtoolsOnConnect: (() -> Void)?

/// Attaches (or clears) the DevTools telemetry sink.
public func fluxDevtoolsSetSink(_ sink: (any VMTelemetrySink)?) {
    fluxDevtoolsSink = sink
}

/// Emits a telemetry event if a DevTools sink is attached.
public func fluxDevtoolsEmit(_ event: TelemetryEvent) {
    fluxDevtoolsSink?.emit(event)
}

/// Encodes a `FluxValue` into `data` per Appendix D §D.5.
func encodeValue(_ value: FluxValue, into data: inout Data) {
    data.append(value.tag)
    switch value {
    case .null:
        break
    case let .int(v):
        data.append(contentsOf: v.bytesLE())
    case let .float(v):
        data.append(contentsOf: v.bitPattern.bytesLE())
    case let .bool(v):
        data.append(v ? 1 : 0)
    case let .str(id):
        data.append(contentsOf: id.bytesLE())
    case let .handlerRef(id):
        data.append(contentsOf: id.bytesLE())
    case let .list(items):
        data.append(contentsOf: UInt16(items.count).bytesLE())
        for item in items {
            encodeValue(item, into: &data)
        }
    case let .record(fields):
        data.append(contentsOf: UInt16(fields.count).bytesLE())
        for (index, field) in fields {
            data.append(contentsOf: index.bytesLE())
            encodeValue(field, into: &data)
        }
    }
}

extension FixedWidthInteger {
    /// Little-endian byte representation.
    func bytesLE() -> [UInt8] {
        withUnsafeBytes(of: self.littleEndian) { Array($0) }
    }
}
